# Working with the T2T reference
Hi, chances are you are working with long-read data if you require the T2T reference (also known as hs1 or chm13v2.0).
Read this paper for more details on how it was made and generated: (note that the most difficult regions were put together manually)
and also that this is one individuals genome and some genomic regions (repeats, retroviral elements etc) may actually give you
a structural variant when variant calling on patient data - T2T is still not the repeat/dark genome standard. For a standardised version, try using 
some of the Human PanGenome "references".
Original Science manuscript for the T2T: https://www.science.org/doi/10.1126/science.abj6987 

T2T files are hosted on NCBI, AWS Human Pangenomics as well as UCSC and are the same, it is up to you which ones you choose.

For downloading any and all of the T2T reference files from NCBI: https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/914/755/GCF_009914755.1_T2T-CHM13v2.0/
For getting any of the NCBI ones in the directory above, you can use wget like this:
wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/101/405/GCF_000101405.1_T2T-CHM13v2.0/GCF_000101405.1_T2T-CHM13v2.0_genomic.fna.gz
(this is an example for the chm13v2.0.fa.gz reference itself which is also in this directory and it should be identical to the AWS one)

The chm13v2.0.fa itself in this folder comes from AWS Human Pangenomics (where the T2T people deposited it):
wget https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=T2T/CHM13/assemblies/analysis_set/ch13v2.0.fa.gz
all other files in this directory starting with GCF_009914755.1 such as RNA annotation (feature tables etc) come from NCBI linked above.

If doing dsbnp and ClinVar, there are liftover databases to the T2T reference genome available at:
https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=T2T/CHM13/assemblies/annotation/liftover/
(I would not recommend doing that, as the T2T reference is not snp-optimized so variant calling misses a lot 
of variants depending on the patient sample - the T2T is European with Neanderthal introgression so if your patient is 
of that descendence, chances are, you will miss things during variant calling and annotation)
# For ClinVar liftover, do:
wget https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=T2T/CHM13/assemblies/annotation/liftover/clinvar.vcf.gz
wget https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=T2T/CHM13/assemblies/annotation/liftover/clinvar.vcf.gz.tbi
# For dbSNP liftover, do:
wget https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=T2T/CHM13/assemblies/annotation/liftover/dbsnp155.vcf.gz
wget https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=T2T/CHM13/assemblies/annotation/liftover/dbsnp155.vcf.gz.tbi

Liftover chain files available at:  https://hgdownload.soe.ucsc.edu/goldenPath/hs1/liftOver/

All of the above and more are also found at the actual github of the T2T reference project: https://github.com/marbl/CHM13 
which contains links to the respective AWS/UCSC/NCBI directories:
1) Sequencing data analysis (fasta files including masked Y/M chromosomes, PARs)
2) Gene annotation (use Refseqv110 liftoff v5.2 which are manually curated, also in this directory - T2T)
3) Repeat annotation (including cytobands, repeat masker, satellites, telomeres etc)
4) Epigenetic profiles (used GIAB HG002)
5) Variant calls lifted over to the T2T - gnomad, 1000 genomes, ClinVar, dbSNP
6) Liftover chain files
7) Non-synthetic regions
8) Papers and citations to all of the above

Generating indexes:
There is a .fai index in this directory which was generated by samtools version 1.21 faidx command
There is a .mmi index in this directory which was generated by the pbmm2 (PacBio minimap2 wrapper) version 1.16.99 index command
There is a T2T_index in this directory which is for STAR alignment of bulk RNA data generated by STAR version 2.7.11a index command
There is a salmon_T2T_index in this directory for getting count pseudoalignment out of bulk RNA generated by salmon version 1.10.3
which uses the GCF_009914755.1_T2T-CHM13v2.0_rna.fna.gz file as "$rna" and keeps duplicates 
(salmon index -t "$rna" -i "$index" --keepDuplicates)

Finally, if you would like to view the T2T reference genome in the IGV on the farm you can do so as an interactive job using X forwarding:
1) when you log in to the farm, you add a -X flag after ssh: ssh -X yourfarmusername@farm22
2) you load IGV: module load igv-2.16.2 
(you can also add this to your .bashrc so it loads automatically after you log in - just nano .bashrc and put module load igv-2.16.2 as a final line)
(you can also add module load ISG/IRODS/1.0 while at it)
3) bsub -I -XF -G yourunixgroup -q yesterday -R 'select[mem>10000] rusage[mem=10000] span[hosts=1]' -M 10000 -n 2 igv.sh 
4) you can then load the hg38 or T2T reference in the interactive IGV

# The hg38 reference
The hg38 reference with decoy regions is hosted at the National Cancer Institute website:
https://gdc.cancer.gov/about-data/gdc-data-processing/gdc-reference-files
while COSMIC is avalible after registration from the online COSMIC database and dbSNP and ClinVar annotations are:
https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0/



